File: ex10_c_dev_kit/examples/continuous_inventory_example.c
/*****************************************************************************
 *                  IMPINJ CONFIDENTIAL AND PROPRIETARY                      *
 *                                                                           *
 * This source code is the property of Impinj, Inc. Your use of this source  *
 * code in whole or in part is subject to your applicable license terms      *
 * from Impinj.                                                              *
 * Contact support@impinj.com for a copy of the applicable Impinj license    *
 * terms.                                                                    *
 *                                                                           *
 * (c) Copyright 2020 - 2021 Impinj, Inc. All rights reserved.               *
 *                                                                           *
 *****************************************************************************/

/*****************************************************************************
 The continuous inventory example below is optimized for faster read rates.
 This example calls continuous_inventory() helper function, which calls
 continuous_inventory() function from the Ex10Reader layer. Meaning the SDK
 is responsible for starting each inventory round, allowing faster read rates.
 For better performance, this example is currently not configured to print each
 inventoried EPC. This can be changed by using the 'verbose' inventory
 configuration parameter.
 The inventory example below is optimized for approximately 256 tags in FOV.
 To adjust dynamic Q algorithm for other tag populations, the following
 parameters should be updated:
 - initial_q
 - max_q
 - min_q
 - min_q_cycles
 - max_queries_since_valid_epc
 For additional details regarding these parameters please refer
 to 'InventoryRoundControl' and 'InventoryRoundControl_2' registers
 descriptions in the Ex10 Reader Chip SDK documentation.
 *****************************************************************************/

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>

#include "board/region.h"
#include "board/time_helpers.h"
#include "ex10_api/application_registers.h"
#include "ex10_api/board_init.h"
#include "ex10_api/ex10_helpers.h"
#include "ex10_api/ex10_reader.h"
#include "ex10_api/rf_mode_definitions.h"
#include "ex10_api/trace.h"


/* Settings used when running this example */
static const uint8_t  antenna                     = 1u;
static const uint16_t rf_mode                     = mode_103;  // original mode_103
static const uint16_t transmit_power_cdbm         = 3000u;
static const uint8_t  initial_q                   = 3u;    //default 8u
static const uint8_t  max_q                       = 15u;
static const uint8_t  min_q                       = 0u;
static const uint8_t  num_min_q_cycles            = 1u;
static const uint16_t max_queries_since_valid_epc = 16u;   // was 16
static const uint8_t  select_all                  = 0u;
static const bool     dual_target                 = true; 
static const uint8_t  session                     = 0u;
static const uint8_t  target                      = 0u;       

// The number of microseconds per second.
const uint32_t us_per_s = 1000000u;

static int continuous_inventory_example(struct Ex10Interfaces ex10_iface,
                                        uint32_t              min_read_rate)
{
    printf("Starting continuous inventory example\n");

    struct InventoryRoundControlFields inventory_config = {
        .initial_q            = initial_q,
        .max_q                = max_q,
        .min_q                = min_q,
        .num_min_q_cycles     = num_min_q_cycles,
        .fixed_q_mode         = false,
        .q_increase_use_query = false,
        .q_decrease_use_query = false,
        .session              = session,
        .select               = select_all,
        .target               = target,
        .halt_on_all_tags     = false,
        .tag_focus_enable     = false,             //  true = read tags once    
        .fast_id_enable       = true,             // false = no TID
    };

    struct InventoryRoundControl_2Fields const inventory_config_2 = {
        .max_queries_since_valid_epc = max_queries_since_valid_epc};

    struct InfoFromPackets packet_info = {0u, 0u, 0u, 0u, {0u}};

    struct ContinuousInventorySummary continuous_inventory_summary = {0};

    struct InventoryHelperParams inventory_params = {
        .antenna               = antenna,
        .rf_mode               = rf_mode,
        .tx_power_cdbm         = transmit_power_cdbm,
        .inventory_config      = &inventory_config,
        .inventory_config_2    = &inventory_config_2,
        .send_selects          = false,
        .remain_on             = false,
        .dual_target           = dual_target,
        .inventory_duration_ms = 0,  // irrelevant for continuous inventory
        .packet_info           = &packet_info,
        .verbose               = false,     //was fakse       
    };

    const struct StopConditions stop_conditions = {
        .max_number_of_tags   = 0u,
        .max_duration_us      = 2u * 1000u * 1000u,          // originally 10
        .max_number_of_rounds = 0u,
    };

    struct ContInventoryHelperParams cihp = {
        .inventory_params = &inventory_params,
        .stop_conditions  = &stop_conditions,
        .summary_packet   = &continuous_inventory_summary};


    if (ex10_iface.helpers->continuous_inventory(&cihp))
    {
        return -1;
    }

    if (continuous_inventory_summary.reason != SRMaxDuration)
    {
        return -1;
    }

    uint32_t const read_rate =
        continuous_inventory_summary.number_of_tags /
        (continuous_inventory_summary.duration_us / us_per_s);

    printf("Read rate = %u - tags: %u / seconds %u.%03u (Mode %u)\n",
           read_rate,
           continuous_inventory_summary.number_of_tags,
           continuous_inventory_summary.duration_us / us_per_s,
           (continuous_inventory_summary.duration_us % us_per_s) / 1000u,
           rf_mode);

    if (continuous_inventory_summary.number_of_tags == 0)
    {
        printf("No tags found in inventory\n");
        return -1;
    }

    if ((min_read_rate) && (read_rate < min_read_rate))
    {
        printf("Read rate of %d below minimal threshold of %d\n",
               read_rate,
               min_read_rate);
        return -1;
    }

    return 0;
}

int main(int argc, char* argv[])
{
    tracepoint(pi_ex10sdk, EXEC_start, __FILE__);
    struct Ex10Interfaces ex10_iface =
        ex10_typical_board_setup(DEFAULT_SPI_CLOCK_HZ, "FCC");         // was "FCC"

    int result = ex10_iface.helpers->check_board_init_status(Application);
    if (result == 0)
    {
        uint32_t min_read_rate = 0;

        if (argc == 2)
        {
            min_read_rate = atoi(argv[1]);
        }

        result = continuous_inventory_example(ex10_iface, min_read_rate);
    }

    ex10_typical_board_teardown();
    tracepoint(pi_ex10sdk, EXEC_end, __FILE__);

    return result;
}

File: ex10_dev_kit/examples/continuous_inventory_example.py
#############################################################################
#                  IMPINJ CONFIDENTIAL AND PROPRIETARY                      #
#                                                                           #
# This source code is the property of Impinj, Inc. Your use of this source  #
# code in whole or in part is subject to your applicable license terms      #
# from Impinj.                                                              #
# Contact support@impinj.com for a copy of the applicable Impinj license    #
# terms.                                                                    #
#                                                                           #
# (c) Copyright 2020 - 2021 Impinj, Inc. All rights reserved.               #
#                                                                           #
#############################################################################
"""
 The continuous inventory example below is optimized for faster read rates.
 This example calls continuous_inventory() helper function, which calls
 continuous_inventory() function from the Ex10Reader layer. Meaning the SDK
 is responsible for starting each inventory round, allowing faster read rates.
 For better performance, this example is currently not configured to print each
 inventoried EPC. This can be changed by using the 'verbose' inventory
 configuration parameter.
 The inventory example below is optimized for approximately 256 tags in FOV.
 To adjust dynamic Q algorithm for other tag populations, the following
 parameters should be updated:
 - INITIAL_Q
 - MAX_Q
 - MIN_Q
 - MIN_Q_CYCLES
 - MAX_QUERIES_SINCE_VALID_EPC
 For additional details regarding these parameters please refer
 to 'InventoryRoundControl' and 'InventoryRoundControl_2' registers
 descriptions in the Ex10 Reader Chip SDK documentation.
"""

from __future__ import (division, absolute_import, print_function,
                        unicode_literals)

import argparse
import threading

from py2c_interface.py2c_python_wrapper import *


# The configuration of the inventory example is set using the definitions below
TRANSMIT_POWER_DBM = 3000           # Transmit power (dBm)
RF_MODE = RfModes.mode_11
REGION = 'FCC'                      # Regulatory region
R807_ANTENNA_PORT = 1               # Which R807 antenna port will be used

INITIAL_Q = 8                       # Q field in the Query command
                                    # (starting Q for dynamic Q algorithm)
MAX_Q = 15                          # Max Q value for dynamic Q
MIN_Q = 0                           # Min Q value for dynamic Q
MIN_Q_CYCLES = 1                    # Number of times we stay on the minimum Q
                                    # before ending the inventory round
MAX_QUERIES_SINCE_VALID_EPC = 16    # Number of queries seen since last good
                                    # EPC before ending the inventory round
SELECT_ALL = 0                      # SELECT field in the Query command
SESSION = 0                         # Session field in the Query command
TARGET = 0                          # Target field in the Query command
DUAL_TARGET = True                  # If True, the Target of subsequent Query
                                    # commands will flip
HALT_ON_ALL_TAGS = False            # If True, the Ex10 will halt on each
                                    # inventoried tag
TAG_FOCUS_ENABLE = False            # Tells a tag to be silent after inventoried
                                    # once.
FAST_ID_ENABLE = False              # Tells a tag to backscatter the TID during
                                    # inventory

pc = c_uint16()
epc = (c_ubyte * 12)()
crc = c_uint16()

packet_info = InfoFromPackets(0, 0, 0,
                              TagReadData(pointer(pc),
                                          ctypes.cast(epc, POINTER(c_ubyte)),
                                          0,
                                          pointer(crc),
                                          None,
                                          0))

continuous_inventory_summary = ContinuousInventorySummary(0, 0, 0, 0, 0, 0, 0)

# pylint: disable=locally-disabled, invalid-name, global-statement
# Indicates whether new packets are available in Ex10Reader
packets_available = threading.Condition()


def run_continuous_inventory_example(min_read_rate=0):
    # pylint: disable=missing-docstring
    print('Starting continuous inventory example')
    try:
        # Init the python to C layer
        py2c = Ex10Py2CWrapper()
        total_singulations = run_continuous_inventory(py2c, INITIAL_Q, min_read_rate)
        if total_singulations <= 0:
            raise RuntimeError('No tags found in inventory')
    finally:
        print('Ending continuous inventory example')
        py2c.ex10_typical_board_teardown()


def run_continuous_inventory(py2c, initial_q, min_read_rate):
    """ Run inventory for the specified amount of time """
    ex10_ifaces = py2c.ex10_typical_board_setup(DEFAULT_SPI_CLOCK_HZ, REGION.encode('ascii'))
    helpers = ex10_ifaces.helpers

    # Put together configurations for the inventory round
    inventory_config = InventoryRoundControlFields()
    inventory_config.initial_q = initial_q
    inventory_config.max_q = MAX_Q
    inventory_config.min_q = MIN_Q
    inventory_config.num_min_q_cycles = MIN_Q_CYCLES
    inventory_config.fixed_q_mode = False
    inventory_config.q_increase_use_query = False
    inventory_config.q_decrease_use_query = False
    inventory_config.session = SESSION
    inventory_config.select = SELECT_ALL
    inventory_config.target = 0
    inventory_config.halt_on_all_tags = HALT_ON_ALL_TAGS
    inventory_config.fast_id_enable = FAST_ID_ENABLE
    inventory_config.tag_focus_enable = TAG_FOCUS_ENABLE

    inventory_config_2 = InventoryRoundControl_2Fields()
    inventory_config_2.max_queries_since_valid_epc = MAX_QUERIES_SINCE_VALID_EPC

    ihp = InventoryHelperParams()
    ihp.antenna               = R807_ANTENNA_PORT
    ihp.rf_mode               = RF_MODE
    ihp.tx_power_dbm          = TRANSMIT_POWER_DBM
    ihp.inventory_config      = pointer(inventory_config)
    ihp.inventory_config_2    = pointer(inventory_config_2)
    ihp.send_selects          = False
    ihp.remain_on             = False
    ihp.dual_target           = DUAL_TARGET
    ihp.inventory_duration_ms = 0   # irrelevant for continuous inventory
    ihp.packet_info           = pointer(packet_info)
    ihp.enforce_gen2_response = False
    ihp.verbose               = False

    stop_conditions = StopConditions()
    stop_conditions.max_number_of_tags   = 0
    stop_conditions.max_duration_us      = 10 * 1000 * 1000
    stop_conditions.max_number_of_rounds = 0

    cihp = ContInventoryHelperParams()
    cihp.inventory_params = pointer(ihp)
    cihp.stop_conditions  = pointer(stop_conditions)
    cihp.summary_packet   = pointer(continuous_inventory_summary)

    assert 0 == helpers.continuous_inventory(pointer(cihp))

    assert 0 < continuous_inventory_summary.reason
    assert StopReason.SRMaxDuration == continuous_inventory_summary.reason
    assert stop_conditions.max_duration_us < continuous_inventory_summary.duration_us

    read_rate = continuous_inventory_summary.number_of_tags/(continuous_inventory_summary.duration_us / 1000000)

    print("Read rate = {} - tags: {} / seconds {:.3f} (Mode {})".format(
          round(read_rate),
          continuous_inventory_summary.number_of_tags,
          continuous_inventory_summary.duration_us / 1000000,
          RF_MODE))

    if min_read_rate > 0 and read_rate < min_read_rate:
        print('Read rate of {} below minimal threshold of {}'.
              format(round(read_rate), min_read_rate))
        assert read_rate >= min_read_rate

    # Cleanup and exit
    ex10_ifaces.reader.stop_transmitting()

    return continuous_inventory_summary.number_of_tags


if __name__ == "__main__":

    parser = argparse.ArgumentParser(description='Continuous inventory example')

    default_read_rate = 0

    parser.add_argument('-r', '--min_read_rate', type=int,
                        default=default_read_rate,
                        help='Minimal read rate threshold'
                             ' , default = {}'.format(default_read_rate))

    args = parser.parse_args()

    run_continuous_inventory_example(args.min_read_rate)
